setuser.8
.TH setuser 8
.SH NAME
setuser \- runs another program under a specified account's environment.
.SH SYNOPSIS
.B setuser
.I account
.I child
.SH DESCRIPTION
.I account
is a single argument.
.I child
consists of one or more arguments. 

.B setuser
sets
.BR $GID ,
.BR $UID ,
.BR $SHELL ,
.BR $HOME ,
and
.BR $USER
to
.IR account 's
gid number, uid number, login shell, home directory, and user name, respectively.
It changes the working directory to
.IR account 's
home, and sets its uid and gid to
.IR account 's
uid and gid, with supplementary groups set up.
It then runs
.IR child .

.B setuser
cannot be run by anyone other than root.
.SH EXIT CODES
.B setuser
exits 111 if it cannot find a UNIX account named
.IB account ,
if it cannot setup the environment, change directory, setgid, setuid, or run
.IR child .
Otherwise its exit code is the same as that of
.IR child .
.SH SEE ALSO
envdir(8),
envini(8),
envuidgid(8),
fghack(8),  
multilog(8),
pgrphack(8),
readproctitle(8),
setlock(8),
setuidgid(8),
softlimit(8),
supervise(8),
svc(8),
svok(8),
svscan(8),
svscanboot(8),
svstat(8),
tai64n(8),
tai64nlocal(8)


CHANGES
Added new program setuser to setuid, setenv, plus chdir before exec.

Added new program envini to read environment variables from an INI file.

Added support for adding a prefix in envdir.

Added "once" flag output to svstat.

Fixed sleeper to do a single atomic write.

Changes in daemontools-encore-1.10
==================================

Added support for fnmatch patterns in multilog.

Increased the multilog file size limit to 2GB.
(Thanks to SATOH Fumiyasu)

Fixed missing separator between system error texts.

Major speed ups in multilog, tai64n, and tai64nlocal.

Added support for signal propagation and disabling setsid behavior in
supervise.  (Thanks to Alan Grow)

Changes in daemontools-encore-1.06
==================================

Added options to svstat to control checking the log script service.

Added -l option to svc to explicitly control the main service.

Fixed race issuing svc commands to log and then main service.

Changed the "safely written" option for multilog to "w".

Fixed daemontools leaking the status file descriptor to its children.

Fixed svc sending commands to both the log and main service.

Changes in daemontools-encore-1.05
==================================

Added options to svup to control checking the log script service.

Eliminated some compiler warnings.

Changes in daemontools-encore-1.04
==================================

Fixed bug in svscan preventing it from starting services with a "log"
script instead of directory.

Changes in daemontools-encore-1.03
==================================

Added -s option to setuidgid to set the supplemental GIDs.
(Thanks to SATOH Fumiyasu)

Changes in daemontools-encore-1.02
==================================

If a supervise is managing both a main and log service, issuing
"svc -dx" on the service will cause the log to stop after the main
service exits.

Changes between daemontools-0.76 and daemontools-encore
=======================================================

Added support to supervise/svc for sending SIGQUIT/USR1/USR2/WINCH.
(original patch by Wayne Marshall, based on patch by Scott Gifford)

Modified supervise to automatically put the forked service into its own
process group to facilitate sending signals to the group.

Added support to supervise/svc for sending signals to the group.

Added support to softlimit for reading limits from environment variables.

Modified supervise to run "start" once before executing "run" to allow
for single-shot initialization, and to run "stop" after "run" exits for
the last time.  The presence of "run" is also optional.

Modified supervise to stop trying to restart the service if it exits 100.
(semantics borrowed from qmail control files)

Added support to multilog for "accustamp" time stamps.
(derived from patch by Matthias Andree)

Added support to supervise for storing the status files in an alternate
directory specified by $SUPERVISEDIR.

Modified multilog to support specifying an alternate "safely written"
filename extension (useful for compressed logs).

Added extended state information to the supervise/status file and svstat.

Added support to supervise for a "notify" script, executed whenever a
subprocess is started or exits.  If present, it is passed the script
name, "start" or "exit", the PID, and the exit code.

Added support for piping to "log" within supervise.  This requires "log"
to be an executable file (instead of a directory as for svscan).

Modified supervise to overwrite status instead of creating a new file.
This allows the "supervise" subdirectory to be non-writeable.

Added a "svup" tool to determine if a service is running.

Modified svscan to optionally redirect all output through a service.
(adapted from http://www.gluelogic.com/code/daemontools/daemontools-0.76-gps-logging-0.01.patch)
Note that this feature works differently from the GlueLogic
modification.  See the svscan(8) man page for details.


setuser.c
#include <sys/types.h>
#include <pwd.h>
#include <unistd.h>
#include "fmt.h"
#include "pathexec.h"
#include "prot.h"
#include "sgetopt.h"
#include "strerr.h"

#define FATAL "setuser: fatal: "

static void usage()
{
  strerr_die1x(100,"setuser: usage: setuser account child");
}

static void nomem()
{
  strerr_die2x(111,FATAL,"out of memory");
}


int main(int argc,const char *const *argv,const char *const *envp)
{
  struct passwd *pw;
  const char *account;
  char strnum[FMT_ULONG];

  account = *++argv;
  if (!account || !*++argv)
    usage();

  pw = getpwnam(account);
  if (!pw)
    strerr_die3x(111,FATAL,"unknown account ",account);

  if (!pathexec_env("HOME",pw->pw_dir)) nomem();
  if (!pathexec_env("SHELL",pw->pw_shell)) nomem();
  if (!pathexec_env("USER",pw->pw_name)) nomem();
  strnum[fmt_ulong(strnum,pw->pw_gid)] = 0;
  if (!pathexec_env("GID",strnum)) nomem();
  strnum[fmt_ulong(strnum,pw->pw_uid)] = 0;
  if (!pathexec_env("UID",strnum)) nomem();

  if (chdir(pw->pw_dir) != 0)
    strerr_die3sys(111,FATAL,"unable to chdir to ", pw->pw_dir);
  if (prot_gid(pw->pw_gid) == -1)
    strerr_die2sys(111,FATAL,"unable to setgid");
  if (prot_gids(pw->pw_name, pw->pw_gid) == -1)
    strerr_die2sys(111,FATAL,"unable to initgroups");
  if (prot_uid(pw->pw_uid) == -1)
    strerr_die2sys(111,FATAL,"unable to setuid");

  pathexec_run(*argv,argv,envp);
  strerr_die3sys(111,FATAL,"unable to run ",*argv);
}
